pkg/runner/test_runner.go

```diff
@@ -6,282 +6,796 @@
+		}
+		tr.targets = rs.Targets
+	}
+	// Initialize remaining fields of the target structures,
+	// build the map and kick off target processing.
+	for _, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		if ts == nil {
+			ts = &targetState{
+				CurPhase: targetStepPhaseInit,
+			}
+		}
+		ts.tgt = tgt
+		ts.resCh = make(chan error)
+		tr.mu.Lock()
+		tr.targets[tgt.ID] = ts
+		tr.mu.Unlock()
+		tr.targetsWg.Add(1)
+		go func() {
+			tr.targetHandler(targetCtx, stepCtx, ts)
+			tr.targetsWg.Done()
+		}()
+	}
+
+	// Run until no more progress can be made.
+	runErr := tr.runMonitor()
+	if runErr != nil {
+		tr.log.Errorf("monitor returned error: %q, canceling", runErr)
+		stepCancel()
+	}
+
+	// Wait for step runners and readers to exit.
+	if err := tr.waitStepRunners(ctx); err != nil {
+		tr.log.Errorf("step runner error: %q, canceling", err)
+		stepCancel()
+		if runErr == nil {
+			runErr = err
+		}
+	}
+	// There will be no more results, reel in all the target runners (if any).
+	tr.log.Debugf("waiting for target runners to finish")
+	targetCancel()
+	tr.targetsWg.Wait()
+
+	// Examine the resulting state.
+	tr.log.Debugf("leaving, err %v, target states:", runErr)
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
+	resumeOk := (runErr == nil)
+	var inFlightTargets []*targetState
+	for i, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		stepErr := tr.steps[ts.CurStep].runErr
+		tr.log.Debugf("  %d %s %v", i, ts, stepErr)
+		if ts.CurPhase == targetStepPhaseRun {
+			inFlightTargets = append(inFlightTargets, ts)
+			if stepErr != statectx.ErrPaused {
+				resumeOk = false
+			}
+		}
+	}
+	tr.log.Debugf("- %d in flight, ok to resume? %t", len(inFlightTargets), resumeOk)
+	tr.log.Debugf("step states:")
+	for i, ss := range tr.steps {
+		tr.log.Debugf("  %d %s %t %t %v", i, ss, ss.stepRunning, ss.readerRunning, ss.runErr)
+	}
+
+	// Is there a useful error to report?
+	if runErr != nil {
+		return nil, runErr
+	}
+
+	// Has the run been canceled?
 	select {
 	case <-ctx.Done():
-		w.log.Debugf("terminate requested while writing target %+v", target)
-	case ch <- target:
-	case <-time.After(timeout):
-		return fmt.Errorf("timeout (%v) while writing target %+v", timeout, target)
+		return nil, statectx.ErrCanceled
+	default:
 	}
-	w.log.Debugf("done writing target %+v, spent %v", target, time.Since(start))
-	return nil
+
+	// Have we been asked to pause? If yes, is it safe to do so?
+	select {
+	case <-ctx.Paused():
+		if !resumeOk {
+			tr.log.Warningf("paused but not ok to resume")
+			break
+		}
+		rs := &resumeStateStruct{
+			Version: resumeStateStructVersion,
+			JobID:   jobID, RunID: runID,
+			Targets: tr.targets,
+		}
+		resumeState, runErr = json.Marshal(rs)
+		if runErr != nil {
+			tr.log.Errorf("unable to serialize the state: %s", runErr)
+		} else {
+			runErr = statectx.ErrPaused
+		}
+	default:
+		// We are not pausing and yet some targets were left in flight.
```

Comment:
Ho do we get here? `runMonitor` has returned, `waitStepRunners` has returned, so I assume all steps have  returned (without an error? otherwise `runErr` would have been not nil and we would have returned before?) How do we get into this state, where we consider steps terminated, monitor terminated, but still targets might be in-flight?

Reply:
that's if plugin "swallowed" a target - it returned from Run with no error but never reported result for a target that was injected.